bytecomp/bytesections.ml

```diff
@@ -53,49 +87,50 @@ let read_toc ic =
     really_input_string ic (String.length Config.exec_magic_number)
   in
   if header <> Config.exec_magic_number then raise Bad_magic_number;
-  seek_in ic (pos_trailer - 8 * num_sections);
-  section_table := [];
+  let toc_pos = (pos_trailer - 8 * num_sections) in
+  seek_in ic toc_pos;
+  let section_table_rev = ref [] in
   for _i = 1 to num_sections do
     let name = really_input_string ic 4 in
     let len = input_binary_int ic in
-    section_table := (name, len) :: !section_table
-  done
+    section_table_rev := (name, len) :: !section_table_rev
+  done;
+  let first_pos, sections =
+    List.fold_left (fun (pos, l) (name, len) ->
+        let section = {name; pos = pos - len; len} in
+        (pos - len, section :: l)) (toc_pos, []) !section_table_rev
+  in
+  { sections; first_pos }
+
+let all t = t.sections
 
-(* Return the current table of contents *)
+let pos_first_section t = t.first_pos
 
-let toc () = List.rev !section_table
+let find_section t name =
+  let rec find = function
+    | [] -> raise Not_found
+    | {name = n; pos; len} :: rest ->
+        if n = name
+        then pos, len
+        else find rest
+  in find t.sections
 
 (* Position ic at the beginning of the section named "name",
    and return the length of that section.  Raise Not_found if no
    such section exists. *)
 
-let seek_section ic name =
-  let rec seek_sec curr_ofs = function
-    [] -> raise Not_found
-  | (n, len) :: rem ->
-      if n = name
-      then begin seek_in ic (curr_ofs - len); len end
-      else seek_sec (curr_ofs - len) rem in
-  seek_sec (in_channel_length ic - 16 - 8 * List.length !section_table)
-           !section_table
+let seek_section t ic name =
+  assert (String.length name = 4);
```

Comment:
I think this assert is no longer necessary now that you use an abstract type (and if you switch to a sum type, it won't be applicable either).

Reply:
removed